### PR TITLE
Solves NQP-146 & NPQ-53 (Initial World Generation & Actor Generation from Pool)

### PR DESCRIPTION
**Pull request by maxilevi**
## Tasks

* [x] World gen system
* [x] Room addition algorithm integrated
* [x] Entity spawning system
* [x] Support for defining pools of entities
* [x] Camera bug fixed
* [x] Dump game map
* [x] Dev view

## How To

How to open the dev view -> F3 key
How to export a map -> gamemap.dump(path)

## Examples

Game GIF

rooms

Dev view GIF

caves2

Example output of the exported dungeon:
```json
{
  "seed": 15,
  "algorithm": "room_addition",
  "width": 50,
  "height": 30,
  "min_room_space": 16,
  "rooms": {
    "0": {
      "area": 15,
      "actors": 1,
      "aspects": 0
    },
    "1": {
      "area": 17,
      "actors": 3,
      "aspects": 0
    },
    "2": {
      "area": 58,
      "actors": 1,
      "aspects": 0
    },
    "3": {
      "area": 44,
      "actors": 3,
      "aspects": 0
    },
    "4": {
      "area": 24,
      "actors": 1,
      "aspects": 0
    },
    "5": {
      "area": 49,
      "actors": 2,
      "aspects": 0
    },
    "6": {
      "area": 41,
      "actors": 2,
      "aspects": 0
    },
    "7": {
      "area": 74,
      "actors": 2,
      "aspects": 0
    },
    "8": {
      "area": 44,
      "actors": 1,
      "aspects": 0
    },
    "9": {
      "area": 66,
      "actors": 2,
      "aspects": 0
    },
    "10": {
      "area": 42,
      "actors": 3,
      "aspects": 0
    },
    "11": {
      "area": 25,
      "actors": 3,
      "aspects": 0
    },
    "12": {
      "area": 29,
      "actors": 2,
      "aspects": 0
    },
    "13": {
      "area": 17,
      "actors": 3,
      "aspects": 0
    },
    "14": {
      "area": 21,
      "actors": 3,
      "aspects": 0
    },
    "15": {
      "area": 15,
      "actors": 1,
      "aspects": 0
    },
    "16": {
      "area": 27,
      "actors": 1,
      "aspects": 0
    },
    "17": {
      "area": 19,
      "actors": 1,
      "aspects": 0
    },
    "18": {
      "area": 15,
      "actors": 1,
      "aspects": 0
    },
    "19": {
      "area": 15,
      "actors": 2,
      "aspects": 0
    }
  }
}
```

### CODE REVIEW

**Comment by Snayff** [comment]
> It is possible to dump a game map to an external file.
> What format? JSON?

Whatever is easiest. JSON works for me. 

> It is possible to generate a map with a "Dev view" (see examples):
> Seperate window or the same game? If so how do you want the UI? How to turn it on? etc

There's an example of how it could look in the item on Youtrack. I think best to close the current UI (camera, message log etc.) And open a new UI element to display the Dec view. As for how to trigger it I am happy to leave it to you. Whatever is simplest. Maybe just on button press?

File: scripts/engine/ui/elements/camera.py [hunk 1]
```diff
@@ -210,6 +213,19 @@ def update_grid(self):
 
         self._draw_grid(tile_positions)
 
+    def _set_initial_position(self, x: int, y: int):
```
Comment (maxilevi) [inline scripts/engine/ui/elements/camera.py:216]: Without doing this the camera wouldn't center on the player if the player was in a different position than around (0, 0)

File: scripts/engine/ui/elements/camera.py [hunk 2]
```diff
@@ -431,15 +447,15 @@ def should_camera_move(self, start_pos: Tuple, target_pos: Tuple) -> bool:
                         return True
                 if edge_end_x > start_x >= edge_end_x - self.edge_size - 2:
                     # player is on the right side, are we moving right?
-                    if 0 < dir_x:
```
Comment (maxilevi) [inline scripts/engine/ui/elements/camera.py:434]: I think it's clearer to have all the variables in the same side

**Comment by Snayff** [comment]
The code looks really different to your previous PRs. Was something different this time? It works, but it's harder to follow than your other commits.

**Comment by maxilevi** [comment]
> The code looks really different to your previous PRs. Was something different this time? It works, but it's harder to follow than your other commits.

Which part specifically?

**Comment by Snayff** [comment]
The reduced frequency of comments, the old style type hints and I think the
terminology is a little different, too, like using cell instead of tile.
It's not a problem, I'm just curious.

**Comment by maxilevi** [comment]
I am not sure actually. I added more comments than usual in the function descriptions so I kind of felt some other comments weren't as necessary, maybe it is that?

> terminology is a little different, too, like using cell instead of tile.

I am not sure why I used that word, since the algorithm uses tile too

> the old style type hints

What are the old style type hints?

**Comment by Snayff** [comment]
By old type hints I meant :returns: in the docstring rather than -> after
the function.

It's interesting to hear your thoughts. Thank you, Maxi.

**Comment by maxilevi** [comment]
> By old type hints I meant :returns: in the docstring rather than -> after the function. It's interesting to hear your thoughts. 

Oh, that wasn't really intentional. I tried to document more the functions so when you generate the documentation with sphinx it would be more complete. However I wasn't aware that adding :returns: to the doc string replaced the -> type hint. I just assumed I wasn't missing any type hints since mypy passed. My bad

**Comment by Snayff** [comment]
No worries. Sphinx certainly does like it, you're right about that. Thanks
again.

**Pull request merged**